packages/compass-sidebar/src/components/sidebar/sidebar.jsx

```diff
@@ -98,21 +100,21 @@ class Sidebar extends PureComponent {
   }
 
   lastExpandedWidth = defaultSidebarWidthOpened;
-  resizableRef = null;
-
-  toggleCollapsed() {
-    if (!this.props.isCollapsed) {
-      this.lastExpandedWidth = boundSidebarWidth(this.resizableRef.size.width);
-
-      this.resizableRef.updateSize({
-        width: sidebarWidthCollapsed,
-        height: '100%'
-      });
-    } else {
-      this.resizableRef.updateSize({
-        width: boundSidebarWidth(this.lastExpandedWidth),
-        height: '100%'
-      });
+
+  toggleCollapsed(updateWidth) {
+    // When a user drags a resizer to expand/collapse we don't want to update
+    // the width as they're controlling it.
```

Comment:
I don't really like passing an argument to control this, but since this component has three separate prop calls it does when it resizes I didn't totally feel comfortable updating the collapsed state in a different function. Might be a nice thing to refactor one day. Open to other suggestions on how we could do these two toggling of the collapsed state flows (one which requires updating width, one where the user is dragging and controlling the width).

Reply:
I might be missing something but as far as I can see it there is little difference between the two cases, they all end up updating `width` to a certain value.

How I see it, you more or less have two React states here that you care about: current `width` and previous `width` (previous one you care about only when toggle button is used), everything else doesn't really matter.

Now we need to consider collapsing logic that goes something like this: when width is less than threshold, it is collapsed, otherwise it is not. As there is nothing but width used here and it's not a very complex computation, this can be derived in render method.

When you look at it like this, collapsed state and the rendered width can be derived just from the width that is stored in React state, additionally `previousWidth` can be stored when toggle is used to collapse the sidebar. This all results in a code kinda like this:

```jsx
updateWidth(width, keepPrevWidth = false) {
  this.setState(({ width: prevWidth }) => ({ width, prevWidth: keepPrevWidth ? prevWidth : null }))
}

...

render() {
  const { width, prevWidth } = this.state;
  const isCollapsed = width < sidebarMinWidthOpened;
  const renderedWidth = isCollapsed ? sidebarWidthCollapsed : boundSidebarWidth(width);

  return (
    <div id="sidebar" style={{ width: renderedWidth }}>
      ...
      <ResizeHandler ... onChange={width => this.updateWidth(width)} />
      <button id="collapse-button" onClick={() => isCollapsed ? this.updateWidth(prevWidth) : this.updateWidth(sidebarWidthCollapsed, true)}>
        ...
  )
}
```

If we want to preserve the global emit, this can be done as an effect (or in `componentDidUpdate` as this component is not a function component) deriving the collapsed state similar to the render logic code, but I can't see it being used for anything in the code except for being stored so maybe we can just drop this logic.